acra-censor/common/logging_logic.go

```diff
@@ -0,0 +1,268 @@
+	if err != nil {
+		writer.logger.WithError(ErrCantReadQueriesFromFileError).WithField(logging.FieldKeyEventCode, logging.EventCodeErrorCensorIOError).Errorln("Can't create QueryWriter instance")
+		writer.logStorage.Close()
+		return nil, err
+	}
+
+	//handling goroutine
+	go func() {
+		for {
+			select {
+			case <-writer.serializationTicker.C:
+				err := writer.dumpBufferedQueries()
+				if err != nil {
+					writer.logger.WithError(ErrComplexSerializationError).WithField(logging.FieldKeyEventCode, logging.EventCodeErrorCensorIOError).Errorln("Can't dump buffered queries")
+				}
+				writer.serializationTicker.Stop()
+				writer.serializationTicker = time.NewTicker(writer.serializationTimeout)
+			case <-signalBackgroundExit:
+				writer.serializationTicker.Stop()
+				err := writer.logStorage.Close()
+				if err != nil {
+					writer.logger.WithError(err).WithField(logging.FieldKeyEventCode, logging.EventCodeErrorCensorIOError).Errorln("Error occurred on exit QueryWriter instance")
+				}
+			case <-signalShutdown:
+				writer.serializationTicker.Stop()
+				err := writer.logStorage.Close()
+				if err != nil {
+					writer.logger.WithError(err).WithField(logging.FieldKeyEventCode, logging.EventCodeErrorCensorIOError).Errorln("Error occurred on shutdown QueryWriter instance")
+				}
+				return
+			default:
+				//do nothing. This means that channel has no data to read yet
+			}
+		}
+	}()
+	return writer, nil
+}
+
+// SetSerializationTimeout sets timeout of dumping captured queries to the file.
+func (queryWriter *QueryWriter) SetSerializationTimeout(timeout time.Duration) {
+	queryWriter.serializationTimeout = timeout
+}
+
+// GetSerializationTimeout gets timeout of dumping captured queries to the file.
+func (queryWriter *QueryWriter) GetSerializationTimeout() time.Duration {
+	return queryWriter.serializationTimeout
+}
+
+// DumpQueries writes all queries into file.
+func (queryWriter *QueryWriter) DumpQueries() error {
+	rawData := queryWriter.serializeQueries(queryWriter.Queries)
+	if err := queryWriter.logStorage.WriteAll(rawData); err != nil {
+		queryWriter.logger.WithError(ErrCantOpenFileError).WithField(logging.FieldKeyEventCode, logging.EventCodeErrorCensorIOError).Errorln("Can't dump queries to storage")
+		return err
+	}
+	return nil
+}
+
+// Release dumps all Captured queries to file and resets Captured queries list.
+func (queryWriter *QueryWriter) Release() {
+	queryWriter.DumpQueries()
+	queryWriter.signalBackgroundExit <- true
+	queryWriter.reset()
+}
+
+func (queryWriter *QueryWriter) reset() {
+	queryWriter.Queries = nil
+}
+
+func (queryWriter *QueryWriter) readStoredQueries() error {
+	queryWriter.mutex.Lock()
+	defer queryWriter.mutex.Unlock()
+
+	q, err := queryWriter.deserializeQueries()
+	if err != nil {
+		queryWriter.logger.WithError(err).WithField(logging.FieldKeyEventCode, logging.EventCodeErrorCensorIOError).Errorln("Can't read stored queries")
+		return err
+	}
+	queryWriter.Queries = q
```

Comment:
we need lock only here. deserialization doesn't use any shared resources but can take some time to read from slow FS and better to not include it into lock time.